### PR TITLE
[Ref/Feat] System Information

### PR DESCRIPTION
**Pull request by CommandMC**
This PR's main goal was to refactor the way we acquire system information. For this, several functions were split off from `src/backend/utils.ts` into their own files/folders, and a few new ones were added
Once that was done, I had the idea to use this information in a dedicated Frontend component, so now we also have a "System Information" tab in the "Settings" category of the sidebar, which allows for easy access of this information (particularly useful if it's needed for support requests)

I've tested this on:
- [X] Linux with a desktop PC & VM
- [X] Windows with a desktop PC & VM
- [X] Steam Deck
- [ ] macOS

Some other highlights include:
- We can now know for certain whether a device is a Steam Deck
- Heroic Version codenames are now kept inside `package.json` for ease-of-access
- A new hook, `useAwaited`, was introduced to easily work with functions returning promises in React. This is particularly useful for IPC

Tests for the various new functions would be neat, but the PR's already getting rather large

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [X] Tested the feature and it's working on a current and clean install.
- [X] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

**Comment by flavioislima** [comment]
Tested here on linux and works pretty nice, will try on macOS as well:
image

```
CPU: 20x 13th Gen Intel(R) Core(TM) i7-13700H
Memory: 33.34 GB (used: 5.91 GB)
GPUs:
  GPU 0:
    Name: NVIDIA Corporation AD106M [GeForce RTX 4070 Max-Q / Mobile]
    IDs: D=2820 V=10de SD=13c0 SV=1462
    Driver: nvidia,  GPU 1:
    Name: Intel Corporation Raptor Lake-P [Iris Xe Graphics]
    IDs: D=a7a0 V=8086 SD=13c0 SV=1462
    Driver: i915
OS: BigLinux based in Manjaro Linux 6.5.0-1-MANJARO (linux)

The current system is not a Steam Deck

Software Versions:
  Heroic: 2.9.1 Boa Hancock
  Legendary: 0.20.32 Dark Energy (hotfix #6)
  gogdl: 0.7.3
  Nile: 1.0.0 Jonathan Joestar
```

**Comment by flavioislima** [comment]
Tested on macOS now:
image

I think that the memory is not very accurate but might be related on how mac reports it?
And it is also slower to get the info on mac than on Linux, just an observation.

**Review by flavioislima — APPROVED** [review]
Looks great.
That's a pretty nice refactor and works pretty well ⚔️

**Comment by CommandMC** [comment]
Memory use is inaccurate on macOS and Linux (there's a `FIXME` about that in the function); Electron's reported value is the *free* amount of RAM, while we should be displaying the *available* amount (which doesn't include disk cache). We'll have to run the `free` command to get that value

The "CPU" card looks a little empty like this, should probably add a logo for Apple

**Review by imLinguin — APPROVED** [review]
Great addition, looks great! MacOS codenames would be nice touch but we can postpone that to be added in the future if you feel like it

File: src/frontend/assets/windows-logo.svg
Comment (arielj) [inline src/frontend/assets/windows-logo.svg:1]: We already have logos that we use for the different systems in the game page and in the library filters. I wonder if we should use those instead for consistency.

We would still need the amd/nvidia/steam ones added

File: src/frontend/screens/Settings/sections/SystemInfo/index.scss [hunk 1]
```diff
@@ -0,0 +1,43 @@
+hr {
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/index.scss:1]: let's add a wrapper to nest all these styles to be sure they don't leak to other components, now these styles are not scoped to this screen and some selectors like `.logo` or `hr` are too generic. something like:

```scss
.systeminfo {
  hr {
  }

  .logo {
  }

  ... etc
}
```

File: src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { ReactComponent as NVIDIALogo } from 'frontend/assets/nvidia-logo.svg'
+import { ReactComponent as IntelLogo } from 'frontend/assets/intel-logo.svg'
+import { ReactComponent as AppleLogo } from 'frontend/assets/apple-logo.svg'
+
+interface Props {
+  model?: string
+}
+
+function VendorLogo({ model }: Props) {
+  if (model?.toLowerCase().includes('nvidia'))
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx:13]:
```suggestion
  if (!model) {
    return <></>
  }

  model = model.lowerCase()

  if (model.includes('nvidia'))
```

Just a suggestion, I think it's simpler to read and slightly faster if we return early if `model` is undefined.

File: src/frontend/screens/Settings/sections/SystemInfo/index.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        <>
+          <h3>
+            {t('settings.navbar.systemInformation', 'System Information')}
+          </h3>
+          <h5>
+            {t(
+              'settings.systemInformation.systemSpecifications',
+              'System Specifications:'
+            )}
+          </h5>
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/index.tsx:108]: Maybe nitpicky here, but we have the h3 and h5 sounding redundant
image

Maybe we can just leave the h3? unless there's a reason to put that h5 there (maybe the idea is to put more content?)

File: public/locales/en/translation.json [hunk 4]
```diff
@@ -622,12 +622,32 @@
...
+            "CPU": "CPU:",
+            "gogdlVersion": "Gogdl: {{gogdl}}",
+            "gpu": "GPU:",
+            "gpuWithNumber": "GPU {{no}}:",
+            "heroicVersion": "Heroic: {{heroic}}",
+            "legendaryVersion": "Legendary: {{legendary}}",
+            "memory": "Memory:",
+            "memoryStats": "{{percentUsed}}% used ({{usedGib}} / {{totalGib}}",
+            "nileVersion": "Nile: {{nile}}",
+            "os": "Software:",
```
Comment (arielj) [inline public/locales/en/translation.json:644]:
```suggestion
            "os": "Operating System:",
```

based on https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/3027/files#diff-15995a3ce6795c647cff923de4a7ae8adbacc9d0c170c74425cb37e319e1e134R37

File: src/frontend/hooks/useAwaited.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+export function useAwaited<T>(getter: () => Promise<T>): T | null
+export function useAwaited<T>(getter: () => Promise<T>, defaultValue: T): T
+export function useAwaited<T>(
+  getter: () => Promise<T>,
+  defaultValue: T | null = null
+): T | null {
+  const [value, setValue] = useState<T | null>(defaultValue)
+
+  // This is `setValue` as long as the component requesting the value is mounted
+  let setValueIfMounted = setValue
```
Comment (arielj) [inline src/frontend/hooks/useAwaited.ts:12]: I think this `let` can be moved inside the `useEffect` so it's not set for every render?

like:

```ts
 useEffect(() => {
    let setValueIfMounted = setValue
    getter().then(setValueIfMounted)
   ...
}
```

File: public/locales/en/translation.json [hunk 6]
```diff
@@ -622,12 +622,32 @@
...
+            "systemInformation": "System Information"
         },
         "open-config-file": "Open Config File",
         "reset-heroic": "Reset Heroic",
         "saves": {
             "warning": "Cloud Saves feature is in Beta, please backup your saves before syncing (in case something goes wrong)"
+        },
+        "systemInformation": {
+            "copyToClipboard": "Copy to clipboard",
+            "CPU": "CPU:",
```
Comment (arielj) [inline public/locales/en/translation.json:635]:
```suggestion
            "cpu": "CPU:",
```

for consistency with the rest (gpu, os)

File: src/frontend/screens/Settings/sections/SystemInfo/cpu.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import VendorLogo from './vendorLogo'
+
+function CPUCard({ cpu }: { cpu: SystemInformation['CPU'] }) {
+  const { model, cores } = cpu
+  const { t } = useTranslation()
+
+  return (
+    <Paper sx={{ padding: 1, height: '100%' }} square>
+      <Typography variant="h6">
+        {t('settings.systemInformation.CPU', 'CPU:')}
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/cpu.tsx:17]:
```suggestion
        {t('settings.systemInformation.cpu', 'CPU:')}
```

for consistency with the other keys (gpu, os)

File: src/frontend/screens/Settings/sections/SystemInfo/gpu.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    deviceId,
+    vendorId,
+    subdeviceId,
+    subvendorId,
+    driverVersion
+  } = gpu
+  const { t } = useTranslation()
+
+  const headingText = showNumber
+    ? t('settings.systemInformation.gpuWithNumber', 'GPU {{no}}:', {
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/gpu.tsx:31]:
```suggestion
    ? t('settings.systemInformation.gpuWithNumber', 'GPU {{number}}:', {
```

Just a suggestion, I think it gives a bit more context to translators

File: src/frontend/screens/Settings/sections/SystemInfo/index.tsx (hunk 3 above)
Comment (CommandMC) [reply src/frontend/screens/Settings/sections/SystemInfo/index.tsx:108]: I originally had clarity in mind with this (your CPU, memory, and GPUs are your *specifications*, while everything on the page in general is *information*), but I'd be fine with removing it as well

**Comment by CommandMC** [comment]
Other than implementing the review comments, I've added f7d042bf787958704a758132e2e917eb13279612 now (since it's a common topic in support requests)

File: src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+function VendorLogo({ model }: Props) {
+  if (!model) return <></>
+
+  model = model.toLowerCase()
+
+  if (model.includes('nvidia'))
+    return <NVIDIALogo className="logo fillWithThemeColor" />
+  if (
+    model?.toLowerCase().includes('amd') ||
+    model?.toLowerCase().includes('advanced micro devices')
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx:23]:
```suggestion
    model.includes('amd') || model.includes('advanced micro devices')
```

File: src/frontend/screens/Settings/sections/SystemInfo/gpu.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      <Grid container spacing={1}>
+        <Grid item xs={2}>
+          <VendorLogo model={vendorString} />
+        </Grid>
+        <Grid item xs={10}>
+          {deviceString}
+          <br />
+          DID={deviceId} VID={vendorId}, DSID={subdeviceId} VSID={subvendorId}
+          <br />
+          Driver: {driverVersion}
```
Comment (arielj) [inline src/frontend/screens/Settings/sections/SystemInfo/gpu.tsx:48]: Maybe we need this to be translated? not sure if we need that for the previous line too

File: src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx (hunk 9 above)
Comment (CommandMC) [reply src/frontend/screens/Settings/sections/SystemInfo/vendorLogo.tsx:23]: Huh, wonder how that happened

**Pull request merged**